pandas/indexes/multi.py

```diff
@@ -867,9 +865,7 @@ def from_arrays(cls, arrays, sortorder=None, names=None):
             if len(arrays[i]) != len(arrays[i - 1]):
                 raise ValueError('all arrays must be same length')
 
-        cats = [Categorical.from_array(arr, ordered=True) for arr in arrays]
-        levels = [c.categories for c in cats]
-        labels = [c.codes for c in cats]
+        labels, levels = zip(*[_factorize(arr) for arr in arrays])
```

Comment:
this same procedure is used elsewhere in the codebase (e.g. in pytables), IIRC. pls do a search. Maybe wrap this in a nicer function (e.g. ``arrays`` -> ``labels``, ``levels``), e.g. what you are doing with ``_factorize``, but even a higher level.